core/src/main/java/com/scalar/db/transaction/consensuscommit/CrudHandler.java

```diff
@@ -148,12 +164,11 @@ private List<Result> scanInternal(Scan scan) throws CrudException {
 
         Snapshot.Key key = new Snapshot.Key(scan, r);
 
-        if (!snapshot.containsKeyInReadSet(key)) {
-          snapshot.put(key, Optional.of(result));
-        }
+        // We always update the read set to create before image using the latest result when the
+        // transaction writes or deletes this record.
```

Comment:
It means that we want to use the latest record state for the before image. We now only use the `readSet` for this purpose.

```java
tx1.get(x)  // x.colA=10
tx2.get(x)
tx2.put(x)  // x.colA=5
tx2.commit()
tx1.scan(x) // x.colA=5
tx1.put(x)  // x.colA=1
tx1.commit()
```

It’s a corner case, but we want to commit tx1 with RCSI.

Reply:
> We always update the read set to create before image by using the latest record (result) because another conflicting transaction might have updated the record after this transaction read it first.

Yeah, it's correct and far better than the previous one.

Reply:
Revised in 9cffae7.